lib/hooks/userspace/hook-grpc.js

```diff
@@ -19,14 +19,73 @@
 var cls = require('../../cls.js');
 var shimmer = require('shimmer');
 var semver = require('semver');
+var SpanData = require('../../span-data.js');
+
 var agent;
 
 var SUPPORTED_VERSIONS = '0.13 - 0.15';
 
-function startBatchWrap(startBatch) {
-  return function startBatchTrace(thing, callback) {
-    // TODO: maybe we only want to do this if a root context exists.
-    return startBatch.call(this, thing, cls.getNamespace().bind(callback));
+function makeClientMethod(useDeprecatedArgumentOrder, method, name) {
+  return function clientMethodTrace() {
+    var root = cls.getRootContext();
+    if (!root) {
+      agent.logger.debug('Untraced gRPC call: ', name);
+      return method.apply(this, arguments);
+    } else if (root === SpanData.nullSpan) {
+      return method.apply(this, arguments);
+    }
+    var span = agent.startSpan('grpc-call-' + name);
+    // Check if the response is through a stream or a callback.
+    if (!method.responseStream) {
+      // Grab the callback which is always required.
+      // Depending on the version of grpc, the position of the callback
+      // function differs.
+      // We need to wrap the callback with the context, to propagate it.
+      var cbIndex;
+      if (useDeprecatedArgumentOrder) {
+        cbIndex = method.requestStream ? 0 : 1;
+      } else {
+        cbIndex = arguments.length - 1;
+      }
+      if (cbIndex >= 0 && cbIndex < arguments.length &&
+          typeof arguments[cbIndex] === 'function') {
+        arguments[cbIndex] = wrapCallback(span, arguments[cbIndex]);
+      }
+    }
+    var call = method.apply(this, arguments);
+    // The user might need the current context in listeners to this stream.
+    cls.getNamespace().bindEmitter(call);
+    if (method.responseStream) {
+      call.on('end', function() {
+        agent.endSpan(span);
+      });
+    }
+    return call;
+  };
+}
+
+/**
+ * Wraps a callback so that the current span for this trace is also ended when
+ * the callback is invoked.
+ * @param {SpanData} span - The span that should end after this callback.
+ * @param {function(?Error, value=)} done - The callback to be wrapped.
+ */
+function wrapCallback(span, done) {
+  var fn = function(err, res) {
+    agent.endSpan(span);
+    done(err, res);
+  };
+  return cls.getNamespace().bind(fn);
+}
+
+function makeClientConstructorWrap(useDeprecatedArgumentOrder,
+                                   makeClientConstructor) {
+  return function makeClientConstructorTrace(methods) {
+    var Client = makeClientConstructor.apply(this, arguments);
+    // If version < 0.14, we need to use the old argument order.
```

Comment:
I am not sure what the comment has to do with the rest of the code here. Perhaps move it below to when you actually check `version_`.